File: src/main/java/edu/cornell/cs/apl/viaduct/imp/informationflow/LabelTerm.java
package edu.cornell.cs.apl.viaduct.imp.informationflow;

import edu.cornell.cs.apl.viaduct.security.FreeDistributiveLattice;
import edu.cornell.cs.apl.viaduct.security.Label;
import edu.cornell.cs.apl.viaduct.security.Principal;
import edu.cornell.cs.apl.viaduct.security.solver.ConstraintValue;

/** A constraint term that stands for an information flow label. */
public abstract class LabelTerm {
  /** Get the label value this term evaluates to. */
  public abstract Label getValue();

  abstract ConstraintValue<FreeDistributiveLattice<Principal>> getConfidentiality();

  abstract ConstraintValue<FreeDistributiveLattice<Principal>> getIntegrity();
}

File: compiler/src/main/kotlin/io/github/apl_cornell/viaduct/security/solver/LabelTerm.kt
package io.github.apl_cornell.viaduct.security.solver

import io.github.apl_cornell.viaduct.algebra.FreeDistributiveLattice
import io.github.apl_cornell.viaduct.algebra.solver.LeftHandTerm
import io.github.apl_cornell.viaduct.algebra.solver.RightHandTerm
import io.github.apl_cornell.viaduct.security.Label
import io.github.apl_cornell.viaduct.security.Principal

/** A symbolic representation of a label expression. */
abstract class LabelTerm {
    /** Returns the value of this term given an assignment of values to every variable in the term. */
    abstract fun getValue(solution: ConstraintSolution): Label

    /** Term that corresponds to performing [Label.confidentiality]. */
    abstract fun confidentiality(): LabelTerm?

    /** Term that corresponds to performing [Label.integrity]. */
    abstract fun integrity(): LabelTerm?

    /** Returns a term representing the confidentiality component. */
    internal abstract val confidentialityComponent: LeftHandTerm<FreeDistributiveLattice<Principal>>

    /** Returns a term representing the integrity component. */
    internal abstract val integrityComponent: RightHandTerm<FreeDistributiveLattice<Principal>>
}
